// ionic-angular/__ivy_ngcc__/es2015/components/tabs/tab-highlight.js
import { Directive, ElementRef } from '@angular/core';
import { DomController } from '../../platform/dom-controller';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../platform/dom-controller';
export class TabHighlight {
    /**
     * @param {?} _elementRef
     * @param {?} _dom
     */
    constructor(_elementRef, _dom) {
        this._elementRef = _elementRef;
        this._dom = _dom;
    }
    /**
     * @param {?} tab
     * @return {?}
     */
    select(tab) {
        if (!tab) {
            return;
        }
        const /** @type {?} */ dom = this._dom;
        dom.read(() => {
            const /** @type {?} */ btnEle = tab.btn.getNativeElement();
            const /** @type {?} */ transform = `translate3d(${btnEle.offsetLeft}px,0,0) scaleX(${btnEle.offsetWidth})`;
            dom.write(() => {
                const /** @type {?} */ ele = this._elementRef.nativeElement;
                ((ele.style))[dom.plt.Css.transform] = transform;
                if (!this._init) {
                    this._init = true;
                    dom.write(() => {
                        ele.classList.add('animate');
                    }, 80);
                }
            });
        }, 32);
    }
}
TabHighlight.ɵfac = function TabHighlight_Factory(t) { return new (t || TabHighlight)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DomController)); };
TabHighlight.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: TabHighlight, selectors: [["", 8, "tab-highlight"]] });
/**
 * @nocollapse
 */
TabHighlight.ctorParameters = () => [
    { type: ElementRef, },
    { type: DomController, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabHighlight, [{
        type: Directive,
        args: [{
                selector: '.tab-highlight'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.DomController }]; }, null); })();
function TabHighlight_tsickle_Closure_declarations() {
    /** @type {?} */
    TabHighlight.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    TabHighlight.ctorParameters;
    /** @type {?} */
    TabHighlight.prototype._init;
    /** @type {?} */
    TabHighlight.prototype._elementRef;
    /** @type {?} */
    TabHighlight.prototype._dom;
}
