// ionic-angular/__ivy_ngcc__/es2015/components/picker/picker-controller.js
import { Injectable } from '@angular/core';
import { App } from '../app/app';
import { Config } from '../../config/config';
import { Picker } from './picker';
/**
 * @hidden
 * \@name PickerController
 * \@description
 *
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../app/app';
import * as ɵngcc2 from '../../config/config';
export class PickerController {
    /**
     * @param {?} _app
     * @param {?} config
     */
    constructor(_app, config) {
        this._app = _app;
        this.config = config;
    }
    /**
     * Open a picker.
     * @param {?=} opts
     * @return {?}
     */
    create(opts = {}) {
        return new Picker(this._app, opts, this.config);
    }
}
PickerController.ɵfac = function PickerController_Factory(t) { return new (t || PickerController)(ɵngcc0.ɵɵinject(ɵngcc1.App), ɵngcc0.ɵɵinject(ɵngcc2.Config)); };
PickerController.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: PickerController, factory: PickerController.ɵfac });
/**
 * @nocollapse
 */
PickerController.ctorParameters = () => [
    { type: App, },
    { type: Config, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PickerController, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.App }, { type: ɵngcc2.Config }]; }, null); })();
function PickerController_tsickle_Closure_declarations() {
    /** @type {?} */
    PickerController.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    PickerController.ctorParameters;
    /** @type {?} */
    PickerController.prototype._app;
    /** @type {?} */
    PickerController.prototype.config;
}
