// ion-image-slide/src/pages/home/home.ts
import { Component, ViewChild } from '@angular/core';
import { NavController, Slides } from 'ionic-angular';

import { SlideController } from "../../components/slide-controller";

@Component({
  selector: 'page-home',
  templateUrl: 'home.html'
})
export class HomePage {

  @ViewChild(Slides) slides: Slides;

  imgSlides: Array<any> = [
    {
      url: 'http://cdn.m-lab.cn/product/58817cb18d6d810058b6fccc.jpg?imageView2/2/w/720'
    },
    {
      url: 'http://cdn.m-lab.cn/product/58817e951b69e6005908ad2f.jpg?imageView2/2/w/720'
    },
    {
      url: 'http://cdn.m-lab.cn/product/58817ea5128fe1006824408f.jpg?imageView2/2/w/720'
    },
    {
      url: 'http://cdn.m-lab.cn/product/58817eb31b69e6005908ade3.jpg?imageView2/2/w/720'
    }
  ];

  constructor(
    public navCtrl: NavController,
    private slideCtrl: SlideController
  ) {

  }

  slideUp(ev: any) {
    let currentIndex = this.slides._activeIndex;
    let slide = this.slideCtrl.create({
      ev: ev,
      list: this.imgSlides,
      index: currentIndex,
      change: this.changeCB.bind(this)
    });
    slide.present({
      ev: {
        data: ev,
        list: this.imgSlides,
        index: currentIndex,
        change: this.changeCB.bind(this)
      }
    });
  }

  changeCB(data: any) {
    this.slides.slideTo(data, 0);
  }

}
